File: Data Structures & Algorithms/rotting-fruit/submission-12.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        // Time and Space Complexity: O(rows * cols)
        set<pair<int, int>> visited;
        set<pair<int, int>> moves{{0,1},{1,0},{0,-1},{-1,0}};
        deque<pair<int, int>> q;
        int fresh{0}, time{0};
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[i].size(); j++) {
                int node{grid[i][j]};
                if (node == 1) fresh++;
                else if (node == 2) {
                    q.push_back({i,j});
                    visited.insert({i,j});
                }
            }
        }
        while (!q.empty() && fresh) {
            int layer{q.size()};
            for (int i = 0; i < layer; i++) {
                auto [r, c] = q.front();
                q.pop_front();
                for (auto [dr, dc] : moves) {
                    int nr{r + dr}, nc{c + dc};
                    if (min(nr, nc) >= 0 && nr < grid.size() && nc < grid[nr].size() && !visited.contains({nr, nc})) {
                        visited.insert({nr, nc});
                        if (grid[nr][nc] == 1) {
                            q.push_back({nr, nc});
                            fresh--;
                        }
                    }
                }
            }
            time++;
        }
        return !fresh ? time : -1;
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-12.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        '''
        Questions:
        4 directionally connected grid? decides move set
        Can I modify the grid in place? Minor space efficiency savings by not maintianing a visited set
        grid is m X n? informs how to itterate through the grid

       Algorithm: Multi Source BFS
        define move set
        visited set
        define q = deque() which stores [[row, col]] of nodes in grid
        distance = 0

        Itterate through grid and add all treasure to q

        While q is not empty:
            layer by layer:
                pop from q
                if node is -1 skip it
                grid[node] = distance

                explore surrounding nodes:
                    if new node is in bounds and is not in visited add to visited
                    if new node is 2147483647 then add to q
            increment distance after each layer
        
        Time and Space Complexity: O(m X n)
        '''
        moves = ((0,1), (1,0), (0,-1), (-1,0))
        visited = set()
        q= deque()
        dist = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 0: 
                    q.append([i,j])
                    visited.add((i,j))
        while q:
            layer = len(q)
            for _ in range(layer):
                r, c = q.popleft()
                grid[r][c] = dist
                for dr, dc in moves:
                    nr, nc = r + dr, c + dc
                    if min(nr, nc) >= 0 and nr < len(grid) and nc < len(grid[nr]) and (nr, nc) not in visited:
                        visited.add((nr, nc))
                        if grid[nr][nc] == 2147483647:
                            q.append([nr, nc])
            dist += 1